Mesh layer with new QGIS Time settings

vcloarec opened the pull request:
![time](https://user-images.githubusercontent.com/7416892/77872855-31e29000-7216-11ea-80fb-82e351dbb4a2.gif)


## Time handling in mesh layer

The time in a mesh layer is defined by :
  - a reference time provided by the data, the project or the user
  - each dataset is associated with a relative times
  - time extent is defined by the first time and the last time of all dataset

```
  Reference time :          AT
  Dataset 1 time            o-----RT------RT-----RT-----------RT
  Dataset 2 time            o---------RT------RT--------RT
  Dataset 3 time            o------------------------------RT-------RT----------RT
  Time extent of layer      o-----<--------------------------------------------->
  
  AT : absolute time (QDateTime)
  RT : relative time (qint64)
```
The class `QgsMeshDataprovidertemporalCapabilities` stores the reference time provided by the data and all the relative times of the dataset. This class has the ability to return dataset index from a dataset group index and a relative time since the reference time.
The reference time (can be different than the provider reference time) and the absolute time extent are stored in the class `QgsMeshTemporalProperties`

The temporal settings in the properties widget are only the reference time and the provider time unit:
![image](https://user-images.githubusercontent.com/7416892/77871643-bf23e580-7212-11ea-9b92-82c906c19c22.png)

The default reference time of the layer is set by (sorted by priority):
- from the data provider if defined in data
- from the project if defined
- from current date with time 00h00
The user can change it if he wants.

## Rendering principle
The 2D and 3D renderers access to the active dataset index from the layer with the time range stored in the `QgsContextRenderer` and in the `Qgs3DMapSettings` :
Relative time is calculated with difference between the time range of `QgsContextRenderer` and the reference time stored by `QgsMeshTemporalProperties`.  The dataset index come from the `QgsMeshDataproviderTemporalCapabilities` that maps the relative time with dataset index.

Then data are brought from the provider with the dataset index.

## Static dataset
There is also the possibility to set static dataset, that is to choose the dataset that will be rendered independently of the QGIS time controller. To do that the user can check the check box that is is at the bottom of the temporal page of the properties widget:
![image](https://user-images.githubusercontent.com/7416892/77871772-20e44f80-7213-11ea-8988-a51ec78c1320.png)
For now, as that was simpler to implement and allows to choose directly wanted dataset, the used can choose independently scalar dataset and vector dataset from combo boxes with the time associated to each dataset (relative time for data without reference time). If wanted, a unique time could be implemented, but that's could lead to inconsistently results with non synchronous dataset.
The static dataset settings are in the temporal page but can be easily put in the source page if wanted.

Gustry committed:
doc about how to compile with 3D, speedup build time with ninja

Samweli committed:
separated wms static range from raster layer temporal properties

Samweli committed:
removed wmst specific logis in temporal tab

nyalldawson committed:
Update vector extents after changing data source for a vector layer

Otherwise the old extents are still used after changing a vector's
source

nyalldawson committed:
[FEATURE] Inbuilt network logging tool

This tool, which is available from the new F12 "dev tools" panel,
is a native port of @rduivenvoorde's network logger plugin.

It shows a list of ongoing and completed network requests, along
with a whole load of useful detail like request and reply status,
headers, errors, SSL configuration errors, timeouts, cache status,
etc.

Also has loads of polish and useful capabilities, such as the ability
to filter requests by URL substrings and status, and you can right-click
requests to open the URL in a browser or copy them as a cURL command.

nyalldawson committed:
Make dev tools margins exactly match layer styling dock

3nids committed:
move tree view from the expression builder to a dedicated class

3nids committed:
remove non working caching of field values

3nids committed:
add init methods for clarity

3nids committed:
remove added method to field formatter registry

3nids committed:
fix typo and dox

3nids committed:
avoid reusing virtual method

Samweli committed:
changed parameters setting in the data source uri

nyalldawson committed:
Fix broken raster temporal properties widget

nyalldawson committed:
Cleanup unused API

nyalldawson committed:
Dialog layout tweaks

nyalldawson committed:
Remove outdated test

nyalldawson committed:
Don't translate null string

nyalldawson committed:
Only show wms-t source settings for wms-t layers

nyalldawson committed:
Correctly use provider's temporal capabilities to set widget initial state and ranges

nyalldawson committed:
QStringLiteral

nyalldawson committed:
Remove duplicate call to triggerRepaint

nyalldawson committed:
Use empty string methods, not ""

nyalldawson committed:
Add a todo

3nids committed:
mark as deprecated

3nids committed:
fix dox

3nids committed:
more dox

nyalldawson committed:
[processing] When running in a batch mode, move the layer adding
actions to the "Autofill" menu

This adds more UI consistency, as currently there's a disconnect
from the autofill options available via the Autofill buttons
vs some options which need to be done from the first rows "..."
button

3nids committed:
add QgsFeatureListComboBox to custom widgets (#35386)

3nids committed:
fix layout of custom widget template

nyalldawson committed:
Don't throw misleading exception

3nids committed:
fix widget name in custom widget

3nids committed:
fix custom widget scripts with a nicer var name

3nids committed:
add missing method

m-kuhn committed:
Improve join dialog layout

nyalldawson committed:
[FEATURE][processing] Port map layer widget wrapper to new c++ api

This is a test run, after merged I'll port the other similar map layer
parameters (feature source, raster layer, vector layer, etc)

3nids committed:
in QgsFeatureFilterModel init the display expression when setting the layer (#35400)

3nids committed:
Merge pull request #35378 from 3nids/qgsexpressiontreeview

move tree view from the expression builder to a dedicated class

m-kuhn committed:
Label cases on join properties

m-kuhn committed:
[joins] Disable caching if editing is enabled

Fixes https://github.com/qgis/QGIS/issues/35403

m-kuhn committed:
Remove "virtual" from cache join label

m-kuhn committed:
Merge pull request #35401 from m-kuhn/join_dlg_layout

Improve join dialog layout

mhugent committed:
Add expanded 0/1 to Layer info in GetProjectSettings response

m-kuhn committed:
Merge pull request #35404 from m-kuhn/join_edtiable_cannot_be_cached

[joins] Disable caching if editing is enabled

mhugent committed:
Adapt expected test output

suricactus committed:
FIX #33355 New Expression to check the edit mode / state

nyalldawson committed:
Indentation

Samweli committed:
removed static wmst settings in provider temporal capabilities

nyalldawson committed:
Use a scoped class for automatic registration/unregistration of dev tool widget factories

nyalldawson committed:
Nicer handling of automatic scroll-to-new-entries mode

Now we only do this if the view was already scrolled to the bottom.

nyalldawson committed:
When pruning logged network requests, never prune expanded out nodes

Otherwise it's rather annoying for users when the request they are
currently examining "magically" vanishes from their screen

nyalldawson committed:
Somedays Travis is just calling out to be punched in the face...

nyalldawson committed:
[processing] Fix file extension for output files is ignored

Fixes #35269

nyalldawson committed:
[feature][processing] Port Feature Source, Raster, Vector and Mesh Layer parameters to new API

Samweli committed:
logic for static temporal range updates when animation is in play

suricactus committed:
Fix unwanted automatic fill of all algorithms when there is no input in the locator

nyalldawson committed:
Default to current layer when appropriate

nyalldawson committed:
Trick autopep

nyalldawson committed:
Fix exception in deprecated classes

nyalldawson committed:
Better repr for QgsProcessingFeatureSourceDefinition

nyalldawson committed:
[processing] Expose extra source options like browsing for layers,
limiting features, and overridding geometry handling in batch mode too

Gustry committed:
update also htmlToMarkdown in the crash report fix #31629

rouault committed:
Save vector layer: for CSV/XLS/XLSX/ODS, only use by default displayed values of a restrict set of widget types

This is a follow-up of https://github.com/qgis/QGIS/pull/35296

This implements the suggestion of
https://github.com/qgis/QGIS/pull/35296#issuecomment-604645943

That is the 'Use displayed value' column is checked by default only when
the field has a widget of type ValueMap, ValueRelation, CheckBox and
RelationReference

And when the Attribute table of the dialog box is not displayed (e.g. when
saving a temporary scratch layer), disable those substitutions.

m-kuhn committed:
[locator] Prefer exact match

Fix https://github.com/qgis/QGIS/issues/35418

gprichter committed:
extend mime-type spec for .qgz

the alternative file ending for **compressed qgis project files**
was missing, causing those files to be opened by an archiver application by default

suricactus committed:
FEAT #27230 #27271 [Aggregates] Allow majority and minority statistics on non-numeric fields

suricactus committed:
Added string MeanLength statistic as an option in the statistical summary widget

suricactus committed:
Documentation dyslexia

m-kuhn committed:
Merge pull request #35415 from suricactus/expr_is_editable

FIX #33355  New Expression to check the edit mode of a layer

suricactus committed:
Optimizations in majority/minority aggregates

rouault committed:
Merge pull request #34233 from Gustry/cmake

doc about how to compile with 3D, speedup build time with ninja

suricactus committed:
Fix string majority/minority determinism

suricactus committed:
Fix forgotten test, string majority/minority now are valid aggregates

suricactus committed:
Updated documentation to clearly state the behavior of aggregate majority/minority

nyalldawson committed:
Fix temporal indicator logic: this indicator should only show for temporal layers
controlled by canvas animation, not other forms of layers-with-static-temporal-properties

nyalldawson committed:
Fix leak, loss of connections to existing temporal properties caused
by creating new object over existing one

nyalldawson committed:
Disable 'automatic' mode for raster layers without data provider temporal capabilities

nyalldawson committed:
Make temporal indicator icon more visible on different themes

nyalldawson committed:
[processing] Multi-selection widget can open inline

Affects enum choices and multi-field selection for now

nyalldawson committed:
GDAL Provider is too debug log chatty

nyalldawson committed:
Make more use of layer type hints to avoid trying to open raster/mesh
layers using the ogr provider

nyalldawson committed:
[processing] Port Raster Band parameter to new widget wrapper API

nyalldawson committed:
[processing] Partial port of Multi Input parameter widget (standard and batch)

PeterPetrik commented on src/3d/qgs3dmapscene.h:
```diff
@@ -114,6 +114,10 @@ class _3D_EXPORT Qgs3DMapScene : public Qt3DCore::QEntity
     void sceneStateChanged();
 
+  public slots:
+    //! update the temporale entities
```
//! Updates the temporal entities

PeterPetrik commented on src/app/3d/qgs3dmaptoolidentify.cpp:
```diff
@@ -87,6 +87,7 @@ void Qgs3DMapToolIdentify::activate()
   }
 
   mCanvas->scene()->registerPickHandler( mPickHandler.get() );
+  mIsActivate = true;
```
mIsActive

PeterPetrik commented on src/app/mesh/qgsmeshlayerproperties.cpp:
```diff
@@ -399,3 +448,33 @@ void QgsMeshLayerProperties::aboutToShowStyleMenu()
   QgsMapLayerStyleGuiUtils::instance()->addStyleManagerActions( m, mMeshLayer );
 }
+
+void QgsMeshLayerProperties::reloadTemporaleProperties()
```
reloadTemporalProperties

PeterPetrik commented on src/app/mesh/qgsmeshlayerproperties.cpp:
```diff
@@ -399,3 +448,33 @@ void QgsMeshLayerProperties::aboutToShowStyleMenu()
+
+void QgsMeshLayerProperties::reloadTemporaleProperties()
+{
+  QgsMeshDataProviderTemporalCapabilities *tempCap = mMeshLayer->dataProvider()->temporalCapabilities();
```
please use full names in variables, no abbreviations:

tempCap  -> temporalCapabilities

PeterPetrik commented on src/app/mesh/qgsmeshlayerproperties.h:
```diff
@@ -68,8 +69,14 @@ class APP_EXPORT QgsMeshLayerProperties : public QgsOptionsDialogBase, private U
     //! Saves a style when appriate button is pressed
     void saveStyleAs();
-
+    //! Prepare style menu
```
Prepares

PeterPetrik commented on src/app/mesh/qgsmeshlayerproperties.h:
```diff
@@ -68,8 +69,14 @@ class APP_EXPORT QgsMeshLayerProperties : public QgsOptionsDialogBase, private U
-
+    //! Prepare style menu
     void aboutToShowStyleMenu();
+    //! reload temporal properties from the provider
```
Reloads

PeterPetrik commented on src/app/mesh/qgsmeshstaticdatasetwidget.cpp:
```diff
@@ -0,0 +1,118 @@
+/***************************************************************************
+    qgsmeshstaticdatasetwidget.cpp
+    -------------------------------------
+    begin                : Marsh 2020
```
March 2020

PeterPetrik commented on src/app/mesh/qgsmeshstaticdatasetwidget.h:
```diff
@@ -0,0 +1,86 @@
+/***************************************************************************
+    qgsmeshstaticdatasetwidget.h
+    -------------------------------------
+    begin                : Marsh 2020
```
March 2020

PeterPetrik commented on src/app/mesh/qgsmeshstaticdatasetwidget.h:
```diff
@@ -0,0 +1,86 @@
+
+
+/**
+ * Itel mist model for dataset continaed in dataset group,
```
Model for datasets contained in the dataset group

PeterPetrik commented on src/app/mesh/qgsmeshstaticdatasetwidget.h:
```diff
@@ -0,0 +1,86 @@
+    //! Synchronizes widgets state with associated mesh layer
+    void syncToLayer();
+
+    //!Applies the settings made in the widget
```
//! Applies

PeterPetrik commented on src/core/mesh/qgsdataset.h:
```diff
@@ -0,0 +1,4 @@
+#ifndef QGSDATASET_H
```
maybe we do not need this file?

vcloarec replied:
sure...

PeterPetrik commented on src/core/mesh/qgsmeshdataprovider.h:
```diff
@@ -802,10 +326,25 @@ class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshDat
+     *
+     *
+     * \param unit the temporal unit
+
```
missing *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovider.h:
```diff
@@ -802,10 +326,25 @@ class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshDat
+    /**
+     * Sets the temporal unit of the provider and reload data if it changes.
+     *
+     *
```
extra line

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+#include "qgsmeshdataprovidertemporalcapabilities.h"
```
missing header (copyright)

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+  const qint64 timeSinceGroupeReference =
+    timeSinceGlobalReference - mGlobalReferenceDateTime.msecsTo( groupReference );
+  const QList<qint64> &datasetTimes = mDatasetTimeSinceGroupReference[group];
+  // No dataset or time too little, return invalid index
```
No dataset or a short time interval

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+  if ( datasetTimes.isEmpty() )
+    return QgsMeshDatasetIndex();
+
+  // If time is less than the time of the first dataset, return the first dataset
```
If requested time is before timestamp of the first dataset...

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+    return QgsMeshDatasetIndex();
+
+  // If time is less than the time of the first dataset, return the first dataset
+  // Allow to display "static" dataset (as terrain elevation) whatever the time
```
remove "whatever the time"

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+      return QgsMeshDatasetIndex( group, i - 1 );
+  }
+
+  // no time found because time too big --> return the last index;
```
The requested time is after last timestamp, return the last dataset

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+  return QgsMeshDatasetIndex( group, datasetTimes.count() - 1 );
+}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndex( int group, double timeSinceGlobalReferenceInHours ) const
```
I am wondering, wouldn't it make sense to use miliseconds everywhere in qgis, so we do need to do 3600 * 1000 * and / 3600 / 1000 here and there? (just a question, maybe more hassle)...

how it is in WMS-T, how the time is stored?

also if we use consistently ms, we do not need to add "InHours" or "inMiliseconds" to all functions, variables

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp:
```diff
@@ -0,0 +1,146 @@
+void QgsMeshDataProviderTemporalCapabilities::addDatasetTime( int group, double time )
+{
+  qint64 unitTimeFactor = 1;
+  switch ( mTemporalUnit )
+  {
+    case QgsUnitTypes::TemporalMilliseconds:
+      unitTimeFactor = 1;
+      break;
+    case QgsUnitTypes::TemporalSeconds:
+      unitTimeFactor = 1000;
+      break;
+    case QgsUnitTypes::TemporalMinutes:
+      unitTimeFactor = 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalHours:
+      unitTimeFactor = 60 * 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalDays:
+      unitTimeFactor = 24 * 60 * 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalWeeks:
+      unitTimeFactor = 7 * 24 * 60 * 60 * 1000;
+      break;
+    default:
+      break;
```
@vcloarec this looks like general utlity maybe needed also by other layer types? I am wondering if it hasn't better place somewhere outside the mesh layer code?

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+#ifndef QGSMESHDATAPROVIDERTEMPORALCAPABILITIES_H
```
missing copyright header

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    QgsMeshDataProviderTemporalCapabilities();
+
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in milliseconds)from the dataset \a group
```
lesser -> less

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in milliseconds)from the dataset \a group
+     *
+     * Returns invalide dataset index if the time is less than the first dataset's time
```
invalide -> invalid

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    QgsMeshDatasetIndex datasetIndexTimeInMilliseconds( int group, qint64 timeSinceGlobalReference ) const;
+
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in hours)from the dataset \a group
```
lesser -> less

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in hours)from the dataset \a group
+     *
+     * Returns invalide dataset index if the time is less than the first dataset's time
```
invalide -> invalid

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Adds a \a reference date/time from a dataset \a group
+     *
+     * \note must be used only by the mesh data provider
```
this is class related to the mesh layers (QgsMeshDataProviderTemporalCapabilities) , so maybe note is not relavent, since it couldn't be used for raster/vectors?

vcloarec replied:
This comment (as for raster) means that this class is done to be handle by the provider and its usage have to be avoided from other class. This is the same comments as for raster provider capabilities.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Adds a \a time (in miliseonds) from a dataset contained in \a group
+     *
+     * \note must be used only by the mesh data provider,
```
this is class related to the mesh layers (QgsMeshDataProviderTemporalCapabilities) , so maybe note is not relavent, since it couldn't be used for raster/vectors?

vcloarec replied:
This comment (as for raster) means that this class is done to be handle by the provider and its usage have to be avoided from other class. This is the same comments as for raster provider capabilities.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    void addGroupReferenceDateTime( int group, const QDateTime &reference ) SIP_SKIP;
+
+    /**
+     * Adds a \a time (in miliseonds) from a dataset contained in \a group
```
miliseconds

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Adds a \a time (in provider unit) from a dataset contained in \a group
+     *
+     * \note must be used only by the mesh data provider,
```
this is class related to the mesh layers (QgsMeshDataProviderTemporalCapabilities) , so maybe note is not relavent, since it couldn't be used for raster/vectors?

vcloarec replied:
This comment (as for raster) means that this class is done to be handle by the provider and its usage have to be avoided from other class. This is the same comments as for raster provider capabilities.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    void addDatasetTime( int group, double  time ) SIP_SKIP;
+
+    /**
+     * Returns true if the data has a reference time
```
Returns whether the reference time is set

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Returns true if the data has a reference time
+     *
```
remove extra *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Returns the reference time
+     *
```
remove extra *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Returns the time extent using the internal reference time
+     * and the first and last times available from the all the dataset
+     *
```
remove extra *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+     * Returns the time extent using \a reference date time
+     * and the first and last times available from the all the dataset
+     */
+    QgsDateTimeRange timeExtent( const QDateTime &reference ) const;
```
For reference time I would create setter (setReferenceTime) and remove this functions. Then `QgsDateTimeRange timeExtent() const;` will return absolute time (and if dataset is relative automatically use reference time)

vcloarec replied:
The principle is here that the provider capabilities has a reference time only if the data provides a reference time. So we couldn't not permit to change this reference time by other part than the data provider.
The reference time coming from user of from project is store in the temporal properties of the layer.
So the `timeExtent(const QDateTime &reference )` is used when the time extent is needed with an external reference time, and `timeExtent()` is used to have the "raw" extent time of the data if data reference time exists

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+     * Returns the temporal unit used to read data
+     *
+     * Temporal units supported are milliseconds, seconds, minutes, hors, days and weeks
+     *
```
remove extra *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Sets the temporal unit (\a temporalUnit) used to read data
+     *
+     * Temporal units supported are milliseconds, seconds, minutes, hors, days and weeks
```
do not state enum here, everyone can click on QgsUnitTypes::TemporalUnit to see options.

vcloarec replied:
hmmm, it is particular:
`QgsUnitTypes::TemporalUnit` support month, year, decade, ...
I have intentionally excluded time unit greater than week because there are not fixed duration time unit depending of the context (ho many day in a month, a year, ...). Those time units are not handle in the implemented conversion.
So, I have add doc about that.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Returns the temporal unit used to read data
+     *
+     * Temporal units supported are milliseconds, seconds, minutes, hors, days and weeks
```
do not state enum here, everyone can click on QgsUnitTypes::TemporalUnit to see options.

vcloarec replied:
hmmm, it is particular:
QgsUnitTypes::TemporalUnit support month, year, decade, ...
I have intentionally excluded time unit greater than week because there are not fixed duration time unit depending of the context (ho many day in a month, a year, ...). Those time units are not handle in the implemented conversion.
So, I have add doc about that.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+    * Clears alls stored reference times and dataset times
+    *
```
extra *

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    * Returns the relative time in milliseconds of the dataset
+    *
+    */
+    qint64 datasetTime( const QgsMeshDatasetIndex &index ) const
```
this function does not seem to me that it should be in the public api, maybe just copy code to the caller?

vcloarec replied:
This function is called by the List model class used to display the list of dataset for static dataset in the properties widget(https://github.com/vcloarec/QGIS/blob/6c6499f723d26d1a052ec12289ea95fd2023488e/src/app/mesh/qgsmeshstaticdatasetwidget.cpp#L112). 
So we need to expose it in the public API, no ?

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+     * Temporal units supported are milliseconds, seconds, minutes, hors, days and weeks
+     *
+     */
+    QgsUnitTypes::TemporalUnit temporalUnit() const;
```
wondering what this temporal unit is? is that the widget data to override the MDAL provider units for the mesh data layer? ? if yes, I think somewhere it would need a doc explaining it. Or is it something else?

vcloarec replied:
yes, it is for override the MDAL provider unit.

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+
+  private:
+
+    //! Holds the globla reference date/time value if exists (min of all groups), otherwise it is invalid
```
globla -> global

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    //! Holds the globla reference date/time value if exists (min of all groups), otherwise it is invalid
+    QDateTime mGlobalReferenceDateTime;
+
+    //! Holds the reference times of each dataset groups
```
Holds the reference time of each dataset groups

PeterPetrik commented on src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:
```diff
@@ -0,0 +1,139 @@
+    //! Holds the reference times of each dataset groups
+    QHash<int, QDateTime> mGroupsReferenceDateTime;
+
+    //! Holds the times of each dataset in milliseconds.
```
Holds the time of each dataset ..

PeterPetrik commented on src/core/mesh/qgsmeshlayer.h:
```diff
@@ -26,6 +26,7 @@
 #include "qgsmeshrenderersettings.h"
 #include "qgsmeshtimesettings.h"
 #include "qgsmeshsimplificationsettings.h"
+#include "qgsmeshlayertemporalproperties.h"
```
either forward declaration or include

PeterPetrik commented on src/core/mesh/qgsmeshlayer.h:
```diff
@@ -34,6 +35,7 @@ class QgsTriangularMesh;
 class QgsRenderContext;
 struct QgsMesh;
 class QgsMesh3dAveragingMethod;
+class QgsMeshLayerTemporalProperties;
```
either forward declaration or include

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.cpp:
```diff
@@ -0,0 +1,82 @@
+
+void QgsMeshLayerTemporalProperties::setDefaultsFromDataProviderTemporalCapabilities( const QgsDataProviderTemporalCapabilities *capabilities )
+{
+  const QgsMeshDataProviderTemporalCapabilities *tempCap = static_cast<const QgsMeshDataProviderTemporalCapabilities *>( capabilities );
```
temporalCapabilities

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.cpp:
```diff
@@ -0,0 +1,82 @@
+
+  if ( mReferenceTime.isValid() )
+    mTimeExtent = tempCap->timeExtent();
+  else // If the provider capabilittyies doesn't provide a reference time, try to bring one frome projet settings
```
capabilittyies -> capabilities

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.cpp:
```diff
@@ -0,0 +1,82 @@
+  else // If the provider capabilittyies doesn't provide a reference time, try to bring one frome projet settings
+  {
+    // Not sure it is a good idea to call instance of project here ...
+    mReferenceTime = QgsProject::instance()->timeSettings()->temporalRange().begin();
```
I think the reference time could be member variable of the QgsMeshDataProviderTemporalCapabilities, and changed by setter when QgsProject::instance()->timeSettings()->temporalRange() is changed or manually changed. probably not good to call instance() here

vcloarec replied:
I think the principle of new temporal classes (capabilities and properties) is :
- capabilities : what can the provider handle from the data, attribute of the class are only set from the provider and from data, as much as possible (provider time unit is a breach in the rule, but it is a provider capability that needs to be used in the capabilities class).
- properties : settings from other part, that is user or project settings.

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.cpp:
```diff
@@ -0,0 +1,82 @@
+}
+
+
+
```
remove extra space

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.h:
```diff
@@ -0,0 +1,72 @@
+#ifndef QGSMESHLAYERTEMPORALPROPERTIES_H
```
missing copyright header

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.h:
```diff
@@ -0,0 +1,72 @@
+ *
+ * \since QGIS 3.14
+ */
+
```
remove extra line

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.h:
```diff
@@ -0,0 +1,72 @@
+ * \since QGIS 3.14
+ */
+
+class CORE_EXPORT QgsMeshLayerTemporalProperties : public QgsMapLayerTemporalProperties
```
what is the difference between QgsMeshLayerTemporalProperties and QgsMeshLayerTemporalCapabilities?

vcloarec replied:
I think the principle of new temporal classes (capabilities and properties) is :
- capabilities : what can the provider handle from the data, attribute of the class are only set from the provider and from data, as much as possible (provider time unit is a breach in the rule, but it is a provider capability that needs to be used in the capabilities class).
- properties : settings from other part, that is user or project settings.

Those classes were also implemented for raster wms-t layer

PeterPetrik commented on src/core/mesh/qgsmeshlayerutils.h:
```diff
@@ -251,17 +251,14 @@ class CORE_EXPORT QgsMeshLayerUtils
-      * \since QGIS 3.12
-      */
-    static QDateTime firstReferenceTime( QgsMeshLayer *meshLayer );
+    static QString formatTime( double hours, const QDateTime &referenceTime, const QgsMeshTimeSettings &settings );
```
Do we need this QgsMeshTimeSettings at all? (these are only for static datasets, right?)...

if this function is only for static dataset and that temporal widgets where it is set, I would remove it from the qgsmeshlayerutils.h and put it in the qgis_app so it is not in the public API

vcloarec replied:
It seems to be used too with mesh calculator for the time format... Needs some refactoring and I wonder if we could not put the time format setting in the QgsSettings?

PeterPetrik replied:
Best to ask Nyall in the official PR

PeterPetrik commented on src/core/mesh/qgsmeshrenderersettings.h:
```diff
@@ -650,21 +650,17 @@ class CORE_EXPORT QgsMeshRendererSettings
+    void setActiveScalarDatasetGroup( int activeScalarDatasetGroup );
+
+    int activeVectorDatasetGroup() const;
+    void setActiveVectorDatasetGroup( int activeVectorDatasetGroup );
```
missing docs

PeterPetrik commented on src/core/mesh/qgsmeshtimesettings.h:
```diff
@@ -53,47 +53,16 @@ class CORE_EXPORT QgsMeshTimeSettings
-
     //! Returns format used for relative time
-    QString relativeTimeFormat() const;
+    QString relativeTimeFormat() const; //not necessary with QGIS time setting
```
I do not understand comment "//not necessary with QGIS time setting" also can you please move comment to the doxy //! comment above?

PeterPetrik commented on src/core/mesh/qgsmeshtimesettings.h:
```diff
@@ -53,47 +53,16 @@ class CORE_EXPORT QgsMeshTimeSettings
-    QDateTime absoluteTimeReferenceTime() const;
-    //! Sets reference time used for absolute time format
-    void setAbsoluteTimeReferenceTime( const QDateTime &absoluteTimeReferenceTime );
+    void setRelativeTimeFormat( const QString &relativeTimeFormat ); //not necessary with QGIS time setting
```
I do not understand comment "//not necessary with QGIS time setting" also can you please move comment to the doxy //! comment above?

PeterPetrik commented on src/gui/qgsmapcanvas.h:
```diff
@@ -133,6 +133,14 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
     void setTemporalController( QgsTemporalController *controller );
 
+    /**
+     * Gets access to the temporal controller that will be used to
+     * update the canvas temporal range.
```
Returns temporal controller

vcloarec replied:
"Gets" is consistent with other docs in the same file

PeterPetrik replied:
ok

mhugent committed:
Merge pull request #35408 from mhugent/server_project_settings_expanded

Server project settings, add 'expanded' attribute

PeterPetrik reviewed (approved):
Thanks for handling this task, the code looks good to me. 

Also before doing official PR, best to rebase to current master, Nyall did some changes for WMS-T so hopefully it does not clash

Can you please fix the small issues I have found and create official PR, so @nyalldawson can take a look and comment on the temporal API ?

mhugent committed:
Revert test gpkg files

nyalldawson committed:
[FEATURE][processing] Port multi-layer input parameter widget to new API

Allows
- expression based inputs
- rearrangement of inputs for multi-layer parameters in models, to set
specific layer orders

nyalldawson committed:
Fix doxygen warning

m-kuhn committed:
Add test for search priority

vcloarec committed:
plug mesh layer to QGIS temporal framework
